File: ports/zig/src/radiotap.zig
//! Radiotap + 802.11 frame parsing.
//! Port of `extract_80211_src_mac_and_rssi`, `freq_to_channel`, `packet_contains_mac`.
const std = @import("std");
const mac = @import("mac.zig");
pub const Mac = mac.Mac;

pub const SrcRssi = struct { mac: Mac, rssi: i8 };

fn le16(p: []const u8, i: usize) u16 {
    return @as(u16, p[i]) | (@as(u16, p[i + 1]) << 8);
}
fn le32(p: []const u8, i: usize) u32 {
    return @as(u32, p[i]) | (@as(u32, p[i + 1]) << 8) |
        (@as(u32, p[i + 2]) << 16) | (@as(u32, p[i + 3]) << 24);
}

/// Map a WiFi frequency (MHz) to a channel number.
pub fn freqToChannel(freq: u32) ?u8 {
    if (freq >= 2412 and freq <= 2484) {
        return @intCast(@divTrunc(@as(i32, @intCast(freq)) - 2407, 5));
    }
    if (freq >= 5000 and freq <= 6000) {
        return @intCast(@divTrunc(@as(i32, @intCast(freq)) - 5000, 5));
    }
    return null;
}

const FieldSpec = struct { size: usize, alignv: usize };

fn fieldSpec(index: usize) ?FieldSpec {
    return switch (index) {
        0 => .{ .size = 8, .alignv = 8 }, // TSFT
        1 => .{ .size = 1, .alignv = 1 }, // Flags
        2 => .{ .size = 1, .alignv = 1 }, // Rate
        3 => .{ .size = 4, .alignv = 2 }, // Channel
        4 => .{ .size = 2, .alignv = 2 }, // FHSS
        5 => .{ .size = 1, .alignv = 1 }, // Antenna signal (RSSI)
        6 => .{ .size = 1, .alignv = 1 }, // Antenna noise
        7 => .{ .size = 2, .alignv = 2 }, // Lock quality
        8 => .{ .size = 2, .alignv = 2 }, // TX power
        9 => .{ .size = 1, .alignv = 1 }, // Antenna
        10 => .{ .size = 1, .alignv = 1 }, // DB antenna signal
        11 => .{ .size = 1, .alignv = 1 }, // DB antenna noise
        12 => .{ .size = 2, .alignv = 2 }, // RX flags
        13 => .{ .size = 2, .alignv = 2 }, // TX flags
        14 => .{ .size = 1, .alignv = 1 }, // RTS retries
        15 => .{ .size = 1, .alignv = 1 }, // Data retries
        else => null,
    };
}

/// Parse a captured frame into (transmitter MAC, RSSI). Null if unparseable.
pub fn extract80211SrcMacAndRssi(data: []const u8) ?SrcRssi {
    if (data.len < 8) return null;
    if (data[0] != 0) return null; // radiotap version

    const radiotap_len: usize = le16(data, 2);
    if (radiotap_len < 8 or radiotap_len > data.len) return null;

    // Collect present bitmaps (chained via bit 31).
    var present_maps: [16]u32 = undefined;
    var map_count: usize = 0;
    var present_offset: usize = 4;
    while (true) {
        if (present_offset + 4 > data.len) return null;
        const word = le32(data, present_offset);
        if (map_count < present_maps.len) {
            present_maps[map_count] = word;
            map_count += 1;
        }
        present_offset += 4;
        if ((word & (@as(u32, 1) << 31)) == 0) break;
    }

    var offset: usize = present_offset;
    var rssi_opt: ?i8 = null;

    var map_idx: usize = 0;
    while (map_idx < map_count) : (map_idx += 1) {
        const present = present_maps[map_idx];
        var bit: usize = 0;
        while (bit < 32) : (bit += 1) {
            if ((present & (@as(u32, 1) << @intCast(bit))) == 0) continue;
            const field_index = map_idx * 32 + bit;
            const spec = fieldSpec(field_index) orelse continue;

            const aligned = if (spec.alignv > 1)
                (offset + (spec.alignv - 1)) & ~(spec.alignv - 1)
            else
                offset;

            if (aligned + spec.size > radiotap_len or aligned + spec.size > data.len)
                return null;

            if (field_index == 5)
                rssi_opt = @bitCast(data[aligned]);

            offset = aligned + spec.size;
        }
    }

    const rssi = rssi_opt orelse return null;

    // 802.11 frame header follows the radiotap header.
    const hdr = data[radiotap_len..];
    if (hdr.len < 24) return null;

    const frame_control = le16(hdr, 0);
    const frame_type = (frame_control >> 2) & 0x3;
    if (frame_type == 1) return null; // control frame

    const to_ds = (frame_control & 0x0100) != 0;
    const from_ds = (frame_control & 0x0200) != 0;

    const min_len: usize = if (to_ds and from_ds) 30 else 24;
    if (hdr.len < min_len) return null;

    const base: usize = if (to_ds and from_ds)
        24 // WDS: addr4
    else if (to_ds)
        10 // STA->AP: addr2
    else if (from_ds)
        16 // AP->STA: addr3
    else
        10; // ad-hoc/mgmt: addr2

    var src: Mac = undefined;
    @memcpy(&src, hdr[base..][0..6]);
    return SrcRssi{ .mac = src, .rssi = rssi };
}

/// True if `target` is the parsed source, or appears verbatim in the raw bytes.
pub fn packetContainsMac(data: []const u8, target: Mac) bool {
    if (extract80211SrcMacAndRssi(data)) |sr| {
        if (std.mem.eql(u8, &sr.mac, &target)) return true;
    }
    if (data.len < 6) return false;
    var i: usize = 0;
    while (i + 6 <= data.len) : (i += 1) {
        if (std.mem.eql(u8, data[i..][0..6], &target)) return true;
    }
    return false;
}

test "freqToChannel" {
    try std.testing.expectEqual(@as(?u8, 1), freqToChannel(2412));
    try std.testing.expectEqual(@as(?u8, 6), freqToChannel(2437));
    try std.testing.expectEqual(@as(?u8, 11), freqToChannel(2462));
    try std.testing.expectEqual(@as(?u8, 36), freqToChannel(5180));
    try std.testing.expectEqual(@as(?u8, null), freqToChannel(1000));
}

test "extract RSSI + source MAC from a synthetic frame" {
    const target = Mac{ 0xDE, 0xAD, 0xBE, 0xEF, 0x00, 0x01 };
    var f = [_]u8{0} ** (9 + 24);
    f[0] = 0; // version
    f[2] = 9; // radiotap length (LE) = 9
    f[4] = 0x20; // present bitmap: bit 5 (antenna signal)
    f[8] = 0xD6; // rssi 0xD6 -> -42
    // 802.11 header at offset 9; addr2 at hdr[10..16]
    var k: usize = 0;
    while (k < 6) : (k += 1) f[9 + 10 + k] = target[k];

    const r = extract80211SrcMacAndRssi(&f).?;
    try std.testing.expectEqual(target, r.mac);
    try std.testing.expectEqual(@as(i8, -42), r.rssi);
    try std.testing.expect(packetContainsMac(&f, target));

    const other = Mac{ 0x11, 0x22, 0x33, 0x44, 0x55, 0x66 };
    try std.testing.expect(!packetContainsMac(&f, other));
    try std.testing.expect(extract80211SrcMacAndRssi(f[0..4]) == null);
}

File: ports/cpp/src/core/radiotap.cpp
#include "radiotap.hpp"

#include <algorithm>
#include <bit>
#include <cstring>
#include <vector>

namespace pf {

static uint16_t le16(std::span<const uint8_t> d, size_t i) {
    return static_cast<uint16_t>(d[i]) | (static_cast<uint16_t>(d[i + 1]) << 8);
}

static uint32_t le32(std::span<const uint8_t> d, size_t i) {
    return static_cast<uint32_t>(d[i]) | (static_cast<uint32_t>(d[i + 1]) << 8) |
           (static_cast<uint32_t>(d[i + 2]) << 16) | (static_cast<uint32_t>(d[i + 3]) << 24);
}

std::optional<uint8_t> freq_to_channel(uint32_t freq) {
    if (freq >= 2412 && freq <= 2484)
        return static_cast<uint8_t>((static_cast<int32_t>(freq) - 2407) / 5);
    if (freq >= 5000 && freq <= 6000)
        return static_cast<uint8_t>((static_cast<int32_t>(freq) - 5000) / 5);
    return std::nullopt;
}

std::optional<std::pair<Mac, int8_t>>
extract_80211_src_mac_and_rssi(std::span<const uint8_t> data) {
    if (data.size() < 8)
        return std::nullopt;
    if (data[0] != 0) // radiotap version
        return std::nullopt;

    size_t radiotap_len = le16(data, 2);
    if (radiotap_len < 8 || radiotap_len > data.size())
        return std::nullopt;

    // Collect present bitmaps (chained via bit 31).
    std::vector<uint32_t> present_maps;
    size_t present_offset = 4;
    while (true) {
        if (present_offset + 4 > data.size())
            return std::nullopt;
        uint32_t word = le32(data, present_offset);
        present_maps.push_back(word);
        present_offset += 4;
        if ((word & (1u << 31)) == 0)
            break;
    }

    size_t offset = present_offset;
    std::optional<int8_t> rssi_opt;

    for (size_t map_idx = 0; map_idx < present_maps.size(); ++map_idx) {
        uint32_t present = present_maps[map_idx];
        for (int bit = 0; bit < 32; ++bit) {
            if ((present & (1u << bit)) == 0)
                continue;
            size_t field_index = map_idx * 32 + static_cast<size_t>(bit);

            size_t size = 0, align = 1;
            switch (field_index) {
                case 0:  size = 8; align = 8; break; // TSFT
                case 1:  size = 1; align = 1; break; // Flags
                case 2:  size = 1; align = 1; break; // Rate
                case 3:  size = 4; align = 2; break; // Channel
                case 4:  size = 2; align = 2; break; // FHSS
                case 5:  size = 1; align = 1; break; // Antenna signal (RSSI)
                case 6:  size = 1; align = 1; break; // Antenna noise
                case 7:  size = 2; align = 2; break; // Lock quality
                case 8:  size = 2; align = 2; break; // TX power
                case 9:  size = 1; align = 1; break; // Antenna
                case 10: size = 1; align = 1; break; // DB antenna signal
                case 11: size = 1; align = 1; break; // DB antenna noise
                case 12: size = 2; align = 2; break; // RX flags
                case 13: size = 2; align = 2; break; // TX flags
                case 14: size = 1; align = 1; break; // RTS retries
                case 15: size = 1; align = 1; break; // Data retries
                default: continue;
            }

            size_t aligned = align > 1 ? ((offset + (align - 1)) & ~(align - 1)) : offset;
            if (aligned + size > radiotap_len || aligned + size > data.size())
                return std::nullopt;

            if (field_index == 5)
                rssi_opt = std::bit_cast<int8_t>(data[aligned]);

            offset = aligned + size;
        }
    }

    if (!rssi_opt)
        return std::nullopt;
    int8_t rssi = *rssi_opt;

    auto hdr = data.subspan(radiotap_len);
    if (hdr.size() < 24)
        return std::nullopt;

    uint16_t frame_control = le16(hdr, 0);
    if (((frame_control >> 2) & 0x3) == 1) // control frame
        return std::nullopt;

    bool to_ds = (frame_control & 0x0100) != 0;
    bool from_ds = (frame_control & 0x0200) != 0;

    size_t min_len = (to_ds && from_ds) ? 30 : 24;
    if (hdr.size() < min_len)
        return std::nullopt;

    size_t base;
    if (to_ds && from_ds)
        base = 24; // WDS: addr4
    else if (to_ds)
        base = 10; // STA->AP: addr2
    else if (from_ds)
        base = 16; // AP->STA: addr3
    else
        base = 10; // ad-hoc/mgmt: addr2

    Mac src{};
    std::ranges::copy(hdr.subspan(base, 6), src.begin());
    return std::make_pair(src, rssi);
}

bool packet_contains_mac(std::span<const uint8_t> data, const Mac& target) {
    if (auto parsed = extract_80211_src_mac_and_rssi(data); parsed && parsed->first == target)
        return true;
    if (data.size() < 6)
        return false;
    for (size_t i = 0; i + 6 <= data.size(); ++i)
        if (std::equal(target.begin(), target.end(), data.begin() + i))
            return true;
    return false;
}

} // namespace pf
